README
TAPS IRC Services - beta
=====================
Core Developers of this release:
 Jacques Wweewer, stinger at ircworld.nl
 Felipe, openglx at starbyte,net


Based on PTlink IRC Services 3, by Lamego and PTlink IRC Software Team .
More information at our web site: http://taps.ircworld.nl/

Building, configuration and running
===================================
Read the 'INSTALL' file.

Supported IRC Servers
=======================
PTlink IRC Server 6


CHANGES
/*
 * CHANGES  - Change log for the current release
 *
 * $Id: CHANGES 33 2010-05-10 02:46:01Z openglx $
 */

  - Fixed issue #27, SVSINFO message always send 0 as peak user count [openglx, 03/Mar/2010]
  - Fixed issue #23, exclude channel register limit for sadmin/soper [thales, 3/Mar/2010] 
  - Fixed issue #34, memoserv read command only accept numbers. Added NEW/ALL
      feature [openglx, 01/Mar/2010]
  - Fixed issue #30, SSET founder on logchannel [thales, 28/Feb/2010] 
  - Fixed issue #21, current user count increases also with stealth user [thales, 28/Feb/2010]
  - Refixed issue #31, working for all Services' pseudo-users now [openglx, 28/Feb/2010]
  - Fixed issue #31, CTCP PING from user to services isn't correctly interpreted [openglx, 15/Dec/2009]
  - Fixed issue #16, nickserv handling modes on services.log [openglx, 12/Sep/2009] 
  - Fixed issue #8, IRCd agnostic
  - Fixed issue #1, creating ev_cs_deop trigger and the "regain op on deop" funcionality
  - Fixed issue #11, adding the notify option on/off for recieving new memo's  
  - BUG #14 fixed, OS EXCEPTION LIST with null who_snid could crash services.
   * TAPS IRC Services 0.0.1 Beta 0, wasn't made publicly available
  - Fixed the issue on the fantasy commands when there is a space after the nick. (This fixes issue #13)
  - Removed all final bits of the never released multichat chatclient
  - Inserted code on the library to support Halfop functions, these where missing
  - Created bs_g_kick module. [openglx, 28/Apr/2008]
  - Created bs_g_voicedevoice module. [openglx, 28/Apr/2008]
  - Changes over changes: bs_g_opme is now bs_g_opdeop, accepting paramenter like !op <target> [openglx, 23/Apr/2008]
  - Added "owner" information at OPERSERV EXCEPTION LIST, changed langfile internally [openglx, 18/Apr/2008]
  - Created bs_g_quote module, to use !quote and send a quote to the channel [openglx, 18/Apr/2008]
  - Default akick expire time can now be configured on the database, this is not hardcoded 
    anymore WHOOOOOOOOOOOO.
  - ChanTriggers are now customizable per bs_g_* module [openglx, 16/Apr/2008]
  - Fixing SVN $Id for all .c and .h files of this release. [openglx, 16/Apr/2008]
  - bs_g_opme (!opme module for BotServ Fantasy) correct and working, needs more tests
    before creating all other things. [openglx, 16/Apr/2008]
  - First edition of the nickserv email blacklist module, which prevents certain
    email addresses to be used on registery for example @yahoo.com
  - Language stuff updated 
  - Increased buffer from 512 to 1024 on ns_getsec, was to small for emails
  - Fixed reference to language (if different) on nickserv identify
  - Nickserv language set is not case sensive anymore
  - Fixed a segmentation fault on the os_exception 
  - Removed all multichat references, we don't have interest on that project now
  - Insert an option to edit inserted quotes instead of delete and re-add

******************************************************************
********* From this point on the TAPS release will start *********
******************************************************************

  PTlink IRC Services 3.10.0 ()

(See PTlink Services 3 Series ChangeLog at ChangeLog/From_PTlink3_Series/ directory).
   


src/l2lh.c
/**********************************************************************
 * PTlink IRC Services is (C) CopyRight PTlink IRC Software 1999-2005 *
 *                     http://software.pt-link.net                    *
 * This program is distributed under GNU Public License               *
 * Please read the file COPYING for copyright information.            *
 **********************************************************************
                                                                                
  Description: .lh file generation tool
                                                                                
 *  $Id: l2lh.c 33 2010-05-10 02:46:01Z openglx $
*/
#include "stdinc.h"
#include "ircservice.h"
#include "lang.h"
#include "ircsvs.h"
#include "strhand.h"

#define MAX_ITEMS	256	/* maximum items on one lang file */
char* lang_items[MAX_ITEMS];
char lang_table[MAX_ITEMS][MAX_LANGS][1024];
static int missing_str[MAX_LANGS];
char* item = NULL;
int itcount = 0;

int find_item(char *name);
int item_posicion(char *name);

/* map \b,\u to the IRC control chars */
static void map2controls(char* line)
{
  char tmp[512];
  char *p = line;
  int i = 0;
  
  while(*p)
    {
      if(*p == '\\')
        {
          if(p[1]=='b' || p[1]=='B') /* bold */
            {
                tmp[i++] = 2;
                p += 2;
                continue;
            }
          if(p[1]=='u' || p[1]=='U') /* underline */
            {
                tmp[i++] = 31;
                p += 2;
                continue;
            }
          if(p[1]=='c' || p[1]=='C') /* color, TBD */
            {
                tmp[i++] = 3;
                p += 2;
                continue;
            }            
        }
      tmp[i++]=*p;      
        ++p;
    }
  tmp[i] = 0;
  strncpy(line, tmp, 512);
}

/* returns a string with format letters */
static char* format(char* str)
{
  static char tmp[64];
  char *p;
  int i = 0;

  p = str;
  while(*p)
    {
      if(*p=='%')
        {
          if(p[1]=='%')
            ++p;
          else
            tmp[i++] = p[1];
        }
      ++p;
    }
  tmp[i]='\0';
  return tmp;
}
/* find an item on the array, if not found add it */
int item_posicion(char *name)
{
  int i = 0;
  while(i<itcount && (strcasecmp(name, lang_items[i])!=0))
    ++i;
  if(i<itcount) /* item was found */
    return i;
  /* we need to add this new item */
  lang_items[i]=strdup(name);
  ++itcount;
  return i;
}

/* find an item on the array */
int find_item(char *name)
{
  int i = 0;
  while(i<itcount && (strcasecmp(name, lang_items[i])!=0))
    ++i;
    
  if(i<itcount) /* item was found */
    return i;
    
  return -1;
}

#define ABORT_PROG \
    fclose(fin);\
    fclose(fout);\
    unlink(outfn);\
    exit(-1);
    
int main(int argc, char* argv[])
{
  FILE *fin, *fout;
  char outfn[256];
  char line[512];
  char *lim1, *lim2;  
  int current_item = 0;
  int current_lang = 0;
  char *c;
  int i, nl;
  char *default_format;
  int linenum = 0;
  
  if(argc!=2)
    {
      printf("Usage: %s file.l\n", argv[0]);
      return -1;
    }
  fin = fopen(argv[1],"rt");
  if(IsNull(fin))
    {
      printf("Could not open %s.\n", argv[1]);
      return -2;
    }
  snprintf(outfn, sizeof(outfn),"%sh",argv[1]);
  fout = fopen(outfn, "w+");
  if(IsNull(fout))
    {
      printf("Could not create %s.\n", outfn);
      return -2;
    } 
  printf("Compiling language file %s\n", argv[1]);
  fprintf(fout, "/* File automatically generated by l2lh\n"
  		" Do not edit this file manually */\n");
  /* start processing the file */
  while(fgets(line,sizeof(line), fin))
    {
      ++linenum;
      if(line[0]=='+') /* its an item mark  */
        {         
          if(line[1]=='+')  /* its a comment, skip it */
            continue;
          lim1 = strchr(&line[1],'(');
          if(IsNull(lim1))
	    {
	       fprintf(stderr,"Missing \"(\" on language item at line %d\n", linenum);
	       ABORT_PROG
	       return -3;
            }
          lim2 = strchr(lim1+1,')');
          if(IsNull(lim2))
	    {
	      fprintf(stderr,"Missing \")\" on language item at line %d\n", linenum);
	      ABORT_PROG
	      return -3;
            }
          (*lim1) = '\0'; ++lim1;
          (*lim2) = '\0';
          if(item) /* close previous item */
            free(item);
          current_item = item_posicion(&line[1]);
          lang2index(lim1, current_lang);  
          if(current_lang == -1)
	    {
	      fprintf(stderr,"Aborting: Unknown language id %s at line %d\n", lim1, linenum);
	      ABORT_PROG
            }
          lang_table[current_item][current_lang][0]='\0';
        }
      else  /* its one item line */
        {
          c = strchr(line,'\n');
          if(c)
            *c='\0';
          c = strchr(line,'\r');
          map2controls(line);
          if(c)
            *c='\0';
          if(line[0]=='@')
          {
            int ipos = find_item(&line[1]);                        
            if(ipos == -1)
            {
              fprintf(stderr,"Unknown reference %s at line %d !\n", line, linenum);
              ABORT_PROG
            }
            else 
            if(ipos == current_item)
            {
              fprintf(stderr,"Invalid self reference %s at line %d !\n", line, linenum);	   
              ABORT_PROG            
            }
            else
            {
              int p;
              strcat(lang_table[current_item][current_lang], lang_table[ipos][current_lang]);
              p = strlen(lang_table[current_item][current_lang]);
              /* we need to remove the trailing "\n" */
              lang_table[current_item][current_lang][p-2] = '\0';
            }
          }
          else if(line[0]=='\0')
            strcat(lang_table[current_item][current_lang], " ");
          else
            strcat(lang_table[current_item][current_lang], line);
          strcat(lang_table[current_item][current_lang], "\\n");
        }
    }
   
  /* Items are now in memory, we are going to create the .lh file*/
  i = 0;  
  while(i<itcount)
    {
      fprintf(fout,"const char* %s[%i] = {\n", lang_items[i], MAX_LANGS);
      if(lang_table[i][0] == NULL || lang_table[i][0][0]=='\0')
        {
          fprintf(stderr,"Missing default lang item for %s\n", 
            lang_items[i]);
          ABORT_PROG
          return 2;
        }
      default_format = strdup(format(lang_table[i][0]));
      for(nl=0; nl < MAX_LANGS; ++nl)
        {
          if(lang_table[i][nl]!=NULL && lang_table[i][nl][0]!='\0')
            {
              if(nl > 0) /* let's validate the format strings */
                {
                  char *l;
                  index2lang(nl, l);
                  /* we dont't check on DATE_FORMATs */
                  if((strstr(lang_items[i], "DATE_FORMAT") == NULL) 
                    && strcmp(default_format, format(lang_table[i][nl])))
                    {
                      fprintf(stderr,"Format string is not correct on %s(%s)\n", 
                        lang_items[i], l);
                        ABORT_PROG
                    }
                }              
              fprintf(fout,"\"%s\"", lang_table[i][nl]);
            }
          else
            {
              fprintf(fout,"\"%s\"", lang_table[i][0]);
              missing_str[nl]++;
            }
          if(nl < MAX_LANGS-1)
            fprintf(fout,",\n");
        }
      free(default_format);
      fprintf(fout,"\n};\n\n");
      ++i;
  }
  fclose(fin);
  fclose(fout);
  /* Lets display some stats */
  printf("Results: ");
  for(nl=0; nl < MAX_LANGS; ++nl)
    {
      char *l;
      index2lang(nl, l);
      if(nl>0)
        printf("; ");
      printf("%s (%3.0f%%)", l, ((itcount- missing_str[nl]) / (float)itcount) * 100);
    }
  printf("\n");
  return 0;
}
